Cloud Recording Near Live Command

jmelancongen opened the pull request:
This adds the `RequestTemporarySegmentDurationOverride` command which allows clients to request that a camera records shorter segments on demand. This allows a client to manage costs by configuring cameras for large segment sizes, but dynamically request shorter segments whenever a end-user needs access to the video that is being recorded right now.

This command is optional, and allows a device to offer a better experience for its cloud recording.

HansBusch commented:
Agree with principle of operation. Please find a shorter name or acronym to keep especially capabilities concise. Below I use ShortSegment but there may be a better term.

Suggest to add an optional element ShortSegment to RecordingConfiguration with elements
* Duration and
* Termination

Suggestion to rename method from Request to Set with clarification that calls overwrite previous requests. This would allow to terminate by setting duration to zero.

Editorial:

  The method should be defined in the recording configuration section.

jmelancongen commented:
Updated based on Hans & F2F feedback:
- Renamed to OverrideSegmentDuration name
- Added the SegmentDurationOverride readonly section in recording configuration target to show what is currently in effect.

bsriramprasad commented on doc/RecordingControl.xml:
```diff
@@ -1470,6 +1470,62 @@ Change Request 2061, 2063, 2065, 2109</revremark>
+        reducing the overall storage costs during normal operation.
+      </para>
+      <para>
+        When the temporary override is active, the device should attempt to close its currently recording segment and then
```
Is there any benefit for client/device to NOT close current recording segment ? Just clarifying as the  requirement is marked as recommended with "**should**" leaving a door open for implementations to keep current segment open?

jmelancongen replied:
As clarified during telco, it's specified as should since it's pretty hard to enforce a behavior, since the device needs a new keyframe to close the segment which can introduce some delays. It's all best effort

bsriramprasad commented on doc/RecordingControl.xml:
```diff
@@ -1470,6 +1470,62 @@ Change Request 2061, 2063, 2065, 2109</revremark>
+        <varlistentry>
+          <term>request</term>
+          <listitem>
+            <para role="param">TargetSegmentDuration [xs:duration]</para>
```
Shouldn't client query segment duration limits (expressed as options?) for client to check what device can accept and allow device to throw a relevant error message if the duration is too small or high?

kieran242 replied:
I agree with this comment and I feel further clarity is needed.

jmelancongen replied:
I'm unsure on that point. This parameter is simply the same as the TargetSegmentDuration of the RecordingConfiguration Target setting. It has the same limits, which is none. I'm not sure why there were no limits documented for the RecordingConfiguration, when Cloud Recording was done, as I was not present then.

Do you know why such limits were not documented or specified at that moment? I'm assuming here that this was discussed, and was voluntarily left open ended. Is it not the case?

jflevesque-genetec replied:
I personally don't remember if it was discussed or not. But if we want to specify a limit now, I suggest to have it done in its own PR since it is unrelated to this new request.

kieran242 replied:
@jmelancongen @jflevesque-genetec I agree.

HansBusch commented on wsdl/ver10/recording.wsdl:
```diff
@@ -683,6 +690,36 @@ IN NO EVENT WILL THE CORPORATION OR ITS MEMBERS OR THEIR AFFILIATES BE LIABLE FO
+					<xs:sequence>
+						<xs:any namespace="##any" processContents="lax" minOccurs="0" maxOccurs="unbounded"/>   <!-- first ONVIF then Vendor -->
+					</xs:sequence>
+					<xs:attribute name="TargetSegmentDuration" type="xs:duration" minOccurs="1">
```
Request root elements must not have xs:any as these result in ugly stubs. For other cases always position the xs:any at the last position as it otherwise eats all subsequent items.

jmelancongen replied:
Fixed

HansBusch commented on wsdl/ver10/recording.wsdl:
```diff
@@ -683,6 +690,36 @@ IN NO EVENT WILL THE CORPORATION OR ITS MEMBERS OR THEIR AFFILIATES BE LIABLE FO
+						<xs:any namespace="##any" processContents="lax" minOccurs="0" maxOccurs="unbounded"/>   <!-- first ONVIF then Vendor -->
+					</xs:sequence>
+				</xs:complexType>
+			</xs:element>
```
same holds for the response. Here e.g. microsoft requires to pass by reference an output parameter.

kieran242 commented:
Reviewed and holding off until updates are made.

HansBusch commented:
Thanks, OverrideSegmentDuration now compiles nicely with Visual Studio :-)

HansBusch commented on wsdl/ver10/schema/onvif.xsd:
```diff
@@ -7759,6 +7759,25 @@ and sample rate. </xs:documentation>
+							</xs:annotation>
+						</xs:element>
+						<xs:element name="Expiration" type="xs:duration" minOccurs="1">
+							<xs:annotation>
```
Expiration should be an xs:dateTime. As it is defined now the value varies on each response and the termination time will jitter depending on the communication delay.

jmelancongen replied:
Initially we were concerned this would jitter based on drift between caller and device times. But I don't feel super strongly on either so I'm fine with changing it if you believe this is clearer for the API

jflevesque-genetec replied:
I don't think the latency or jitter between the client and server has any critical impact in this scenario. Even if it took 5s for the device to receive the command, the device will at most do short file durations for the duration specified in the request.

HansBusch commented:
Question regarding behavior over reboot. When clients use short expiration duration they have to frequently update it and storing in flash is problematic. Can we add a note like ```duration of expiration is expected to be in the range of seconds or minutes and may not persist after device restart.``` ?

HansBusch commented:
Question, what behavior is expected when a client changes the recording configuration and an override is active? For the sake of simplicity I would prefer that this deletes any previous override.

HansBusch commented on wsdl/ver10/recording.wsdl:
```diff
@@ -683,6 +690,32 @@ IN NO EVENT WILL THE CORPORATION OR ITS MEMBERS OR THEIR AFFILIATES BE LIABLE FO
 				<xs:anyAttribute processContents="lax"/>
 			</xs:complexType>
+			<xs:element name="OverrideSegmentDurationRequest">
+				<xs:complexType>
```
Syntax error: the request name must not include suffix "Request". Needs to change also below in the message definition.

jmelancongen commented:
Updated the element name without Request. SOAP conventions are pretty unclear to the uninitiated!

I agree with both statements regarding reboots & configuration upgrade. I've added a mention to those in the spec. I did not add the range part though, as we already added following the F2F that "The override duration shall not exceed 1 hour." in the spec. This seems like a similar, but stronger requirement so I kept that one.

kieran242 commented:
Already approved. Re-Reviewed updates and I am happy. Good catch by @HansBusch and you have already fixed it by the time I had re-reviewed.
